pkg/storage/redis/redis.go

```diff
@@ -208,6 +210,50 @@ func (db *DB) ClearDeleted(_ context.Context, cutoff time.Time) {
 	}
 }
 
+// Sync returns a channel to the caller, when there is a change to the version set,
+// sending message to the channel to notify the caller.
+func (db *DB) Sync(ctx context.Context) chan struct{} {
+	ch := make(chan struct{})
+	go func() {
+		c := db.pool.Get()
+		defer func() {
+			c.Close()
+			close(ch)
+		}()
+
+		// Setup notifications, we only care about changes to db.version_set.
+		if _, err := c.Do("CONFIG", "SET", "notify-keyspace-events", "Kz"); err != nil {
+			log.Error().Err(err).Msg("failed to setup redis notification")
+			return
+		}
+
+		psc := redis.PubSubConn{Conn: c}
+		if err := psc.PSubscribe("__keyspace*__:" + db.versionSet); err != nil {
+			log.Error().Err(err).Msg("failed to subscribe to version set channel")
+			return
+		}
+		for {
+			select {
+			case <-ctx.Done():
+				log.Error().Err(ctx.Err()).Msg("stop subscribing to version set channel")
+				return
+			default:
+			}
+			switch v := psc.Receive().(type) {
+			case redis.Message:
+				log.Debug().Str("action", string(v.Data)).Msg("Got redis message")
+				if string(v.Data) == watchAction {
+					ch <- struct{}{}
```

Comment:
this can block leaving an orphaned goroutine. We should do a select with ctx.Done too:

```
select {
case <-ctx.Done(): // ...
case ch <-struct{}{}: // ...
}
```

Reply:
Fixed